airflow-core/src/airflow/ui/src/layouts/Details/Grid/Grid.tsx

```diff
@@ -77,6 +90,46 @@ export const Grid = ({ dagRunState, limit, runType, showGantt, triggeringUser }:
     triggeringUser,
   });
 
+  const selectedVersion = useSelectedVersion();
+
+  // Fetch lineage-filtered structure when filter is active
+  const { data: lineageStructure } = useStructureServiceStructureData(
+    {
+      dagId,
+      externalDependencies: false,
+      includeDownstream: filterMode === "downstream" || filterMode === "both",
+      includeUpstream: filterMode === "upstream" || filterMode === "both",
+      root: filterMode !== "none" && filterRoot !== undefined ? filterRoot : undefined,
+      versionNumber: selectedVersion,
+    },
+    undefined,
+    {
+      enabled: selectedVersion !== undefined && filterMode !== "none" && filterRoot !== undefined,
+    },
+  );
+
+  // Extract allowed task IDs from lineage structure when filter is active
+  const allowedTaskIds = useMemo(() => {
+    if (filterMode === "none" || filterRoot === undefined || lineageStructure === undefined) {
+      return undefined;
+    }
+
+    const taskIds = new Set<string>();
+
+    lineageStructure.nodes.forEach((node) => {
+      const addNodeAndChildren = (currentNode: typeof node) => {
+        taskIds.add(currentNode.id);
+        if (currentNode.children) {
+          currentNode.children.forEach((child) => addNodeAndChildren(child));
+        }
+      };
+
+      addNodeAndChildren(node);
```

Comment:
If we want to make this a function, then we should declare it outside of the forEach loop.